[concurrency] implement the 'unsafe' option for @actorIndependent

kavon opened the pull request:
This PR implements `@actorIndependent(unsafe)`, which disables static safety checks for accesses to actor-isolated storage. It is otherwise semantically the same as `@actorIndependent` (without the option) that is the 'safe' version.

TODOs:
- [x] Implement parsing for this attribute.
- [x] Conditionally disable safety checks in the type checker.
- [x] ~Fix missing detection of mutable state access from `@actorIndependent` computed property~, Reported as SR-13735, but already fixed in `main`.
- [x] Add `@actorIndependent(unsafe)` regression tests.
- [x] Add serialization / deserialization regression tests.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

rjmccall commented on include/swift/AST/Attr.def:
```diff
@@ -578,6 +578,12 @@ SIMPLE_DECL_ATTR(actorIndependent, ActorIndependent,
+  OnClass | OnFunc | OnVar | OnSubscript | ConcurrencyOnly |
+  ABIStableToAdd | ABIStableToRemove |
+  APIStableToAdd | APIBreakingToRemove,
+  104)
```
This all looks right.

kavon committed:
fix diagnostic messages that said '@actorIsolated' for @actorIndependent

kavon committed:
implemented parsing and typechecking for @actorIndependent(unsafe)

[broken] first impl of @actorIndependent in the type checker.

[broken] fixed mistake in my parsing code wrt invalid source range

[broken] found another spot where ActorIndependent needs custom handling

[broken] incomplete set of @actorIndependent(unsafe) tests

updates to ActorIndependentUnsafe

[fixed] add FIXME plus simple handling of IndependentUnsafe context

finished @actorIndependent(unsafe) regression tests

added wip serialization / deserialization test

focus test to just one actor class

round-trip serialize/deserialize test for @actorIndependent

serialize -> deserialize -> serialize -> compare to original

most of doug's comments

addressed robert's comments

fix printing bug; add module printing to regression test

[nfc] update comment for ActorIsolation::IndependentUnsafe

rjmccall reviewed (commented):
Looking good

kavon committed:
add regression tests for SR-13735

DougGregor commented on include/swift/AST/Attr.h:
```diff
@@ -1329,6 +1333,28 @@ class ReferenceOwnershipAttr : public DeclAttribute {
+    Bits.ActorIndependentAttr.kind = unsigned(kind);
+  }
+
+  ActorIndependentAttr(ActorIndependentKind kind)
```
You can remove this constructor by adding an `= false` to the `isImplicit` parameter of the subsequent constructor.

DougGregor commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -4207,22 +4207,22 @@ ERROR(global_actor_isolated_requirement_witness_conflict,none,
       ())
-ERROR(actorisolated_not_actor_member,none,
-      "'@actorIsolated' can only be applied to actor members and "
+ERROR(actorindependent_not_actor_member,none,
```
Thank you for fixing my bugs ;)

DougGregor commented on lib/Serialization/ModuleFormat.h:
```diff
@@ -1807,6 +1807,11 @@ namespace decls_block {
 
+  using ActorIndependentDeclAttrLayout = BCRecordLayout<
+    ActorIndependent_DECL_ATTR,
+    BCFixed<2>  // safety value
```
This only needs one bit?

DougGregor commented on test/Serialization/attr-actorindependent.swift:
```diff
@@ -0,0 +1,32 @@
+
+// CHECK-NOT: UnknownCode
+// CHECK: <ActorIndependent_DECL_ATTR abbrevid=190 op0=1/>
+// CHECK: <ActorIndependent_DECL_ATTR abbrevid=190 op0=0/>
```
We might that the abbrevid value will be brittle. You could replace the `190` with `{{[0-9]+}}`, because we don't really care what the value is.

DougGregor commented on test/Serialization/attr-actorindependent.swift:
```diff
@@ -0,0 +1,32 @@
+// RUN: %empty-directory(%t)
+// RUN: %target-swift-frontend -enable-experimental-concurrency -emit-module -o %t/a.swiftmodule -primary-file %s -module-name main
+// RUN: llvm-bcanalyzer -dump %t/a.swiftmodule | %FileCheck %s
```
An addition check you might add is to use `%target-swift-ide-test` to print out `%t/a.swiftmodule` and use FileCheck to ensure that you get `@actorIndependent` and `@actorIndependent(unsafe)` where you expect it. That'll test the round-tripping aspect.

DougGregor reviewed (approved):
This is looking good! Just a few minor requests.

CodaFi commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -286,19 +286,27 @@ class AttributeChecker : public AttributeVisitor<AttributeChecker> {
       if (var->hasStorage()) {
-        diagnoseAndRemoveAttr(attr, diag::actorisolated_mutable_storage);
-        return;
+        switch(attr->getKind()) {
```
Nit: llvm-style switches please.

kavon replied:
What would that look like? I wasn't able to find anything on the web.

CodaFi replied:
Huh, that's an interesting oversight in the coding standards then. Anyhow

```
switch (subject) {
case <case>:
  <stmt>
...
}
```

`clang-format` can do this for you.

CodaFi commented on lib/AST/Attr.cpp:
```diff
@@ -776,6 +776,22 @@ bool DeclAttribute::printImpl(ASTPrinter &Printer, const PrintOptions &Options,
+    // possible suffix
+    auto aiAttr = static_cast<const ActorIndependentAttr*>(this);
+    switch (aiAttr->getKind()) {
+      case ActorIndependentKind::Safe: break;
```
And here

CodaFi commented on include/swift/AST/ActorIsolation.h:
```diff
@@ -70,8 +73,19 @@ class ActorIsolation {
+  static ActorIsolation forIndependent(ActorIndependentKind indepKind) {
+    ActorIsolation::Kind isoKind;
+    switch (indepKind) {
+      case ActorIndependentKind::Safe: {
```
And here

CodaFi commented on lib/Serialization/ModuleFormat.h:
```diff
@@ -1807,6 +1807,11 @@ namespace decls_block {
     BCFixed<2>  // inline value
   >;
 
+  using ActorIndependentDeclAttrLayout = BCRecordLayout<
```
Please bump the module format version.

DougGregor commented:
The code's looking good. One last request (sorry): could you do a rebase to clean up the history a little bit?

kavon commented:
Ok I cleaned it up into 3 commits. In future PR's I'll go with a rebasing strategy.

DougGregor commented:
@swift-ci smoke test

atrick commented on include/swift/AST/ActorIsolation.h:
```diff
@@ -48,6 +48,9 @@ class ActorIsolation {
     Independent,
+    /// Semantically the same as Independent, but drops the static safety check
+    /// that prevents the use of actor-independent mutable state.
+    IndependentUnsafe,
```
If `Independent` and `IndependentUnsafe` have the same semantics, then a well-defined program will have identical behavior in both cases. If that's the case, then this should be called `IndependentUnchecked`.

However, if this attribute is meant to allow state to be accessed across actors, leaving it up to the programmer to protect that state via synchronization or same-thread requirements, then `IndependentUnsafe` is the correct term, but it definitely does not have the same semantics.

kavon replied:
The latter is intended. I'll reword the comment.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
Let's go ahead and merge, and we can tweak later if needed.
